tools/descriptor_update.py

```diff
@@ -0,0 +1,137 @@
+        match = match[1:-1]
+
+    return new.format(match)
+
+
+def bracketed_replacer_nolegacy(new, match_obj):
+    match = match_obj.group(1)
+    if "LegacyDescr" in match or match.startswith("l"):
+        # Already converted to explicit LegacyDescriptor use.
+        return match_obj.group(0)
+    if match == "ts":
+        # Tokenizer uses ts->fields and ts->field_size, scatter shot.
+        return match_obj.group(0)
+
+    print(match_obj, match)
+    if match[0] == "(" and match[-1] == ")":
+        match = match[1:-1]
+
+    return new.format(match)
+
+
+def apply_arrfunc_fix(content):
+    """
+    The arrfunc `->f` slot should be accessed via the dtype struct.
+    """
+    match = rf"({EXPRESSION})->f->"
+    return re.sub(match, r"PyDataType_ARRFUNCS(\1).", content)
+
+
+def elsize_fix(content, *, is_public):
+    """
+    The ->elsize is moved and needs special access.
+    """
+    # Sometimes `PyArray_DESCR(...)->elsize` is used, but we have a convenience
+    # macro for that:
+
+    match = rf"PyArray_DESCR\(({EXPRESSION})\)->elsize"
+    content = re.sub(match, partial(bracketed_replacer, "PyArray_ITEMSIZE({})"), content)
+
+    if not is_public:
+        # If we have a descriptor rename elsize to `->itemsize`
+        match = rf"({EXPRESSION})->elsize"
+        content = re.sub(match, r"\1->itemsize", content)
+    else:
+        # If public API rename to a new `PyDataType_ITEMSIZE` to allow
+        # NumPy version dependent logic:
+        match = rf"({EXPRESSION})->elsize"
+        content = re.sub(match, partial(bracketed_replacer, "PyDataType_ITEMSIZE({})"), content)
+
+    return content
+
+
+def simple_fix(content, *, name, apply_filter=True):
+    # See elsize fix, no rename so no change internally.
+    if apply_filter:
+        # If public API rename to a new `PyDataType_ITEMSIZE` to allow
+        # NumPy version dependent logic:
+        match = rf"({EXPRESSION})->{name}"
+        MACRO_NAME = f"PyDataType_{name.upper()}"
+        content = re.sub(match, partial(bracketed_replacer_nolegacy, MACRO_NAME + "({})"), content)
+
+    return content
+
+
+
+if __name__ == "__main__":
+
+    for file in files_iter():
+        # Public C-API sometimes needs a different change
+        # (or rather the public way is more cumbersome)
+        is_public = "include/numpy" in str(file) or "f2py" in str(file)
+        content = file.read_text()
+
+        # Apply the fixers:
+        content = simple_fix(content, name="names")
+        content = simple_fix(content, name="fields")
+        content = simple_fix(content, name="subarray")
+        content = simple_fix(content, name="metadata")
+        content = simple_fix(content, name="c_metadata")
+
+        #content = apply_arrfunc_fix(content)
+        #content = elsize_fix(content, is_public=is_public)
+        #content = simple_fix(content, apply_filter=is_public, name="alignment")
+
+        # Write the result:
+        file.write_text(content)
```

Comment:
Do you see this being useful once this PR is merged? If so, there are some linting problems. If not, maybe we don't need it in the repo?

Reply:
Not particularly,  I thought it might be intersting for reviewers to see.  Let me remove it, we can add it back in a dedicated step if anyone thinks its useful.